File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/owners/transfer/TransferOwnershipViewModel.kt
@file:Suppress("PackageNaming", "MagicNumber", "TooManyFunctions")

package app.pantopus.android.ui.screens.homes.owners.transfer

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.homes.TransferOwnerRequest
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.homes.HomeOwnersRepository
import app.pantopus.android.ui.components.PantopusFieldState
import app.pantopus.android.ui.screens.homes.owners.transfer.components.ConfirmSheetParty
import app.pantopus.android.ui.screens.homes.owners.transfer.components.SplitSegment
import app.pantopus.android.ui.screens.shared.form.FormFieldState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/** Nav-arg key for the home id consumed via [SavedStateHandle]. */
const val TRANSFER_HOME_ID_KEY = "homeId"

/** Tone + text payload the form turns into a transient toast. */
data class TransferToast(
    val text: String,
    val isError: Boolean,
)

/** Visibility states for the bottom confirmation sheet. */
enum class ConfirmSheetPhase {
    Hidden,
    Visible,
    Authenticating,
    Dismissing,
}

/**
 * Aggregate UI state for the Transfer Ownership form. Mirrors the iOS
 * `TransferOwnershipViewModel` projection so snapshot / VM tests align.
 */
data class TransferOwnershipUiState(
    val homeContext: TransferOwnershipSampleData.HomeContext,
    val recipient: TransferOwnershipSampleData.RecipientSeed,
    val currentUser: TransferOwnershipSampleData.OwnerSeed,
    val coOwners: List<TransferOwnershipSampleData.OwnerSeed>,
    val recipientIsBackendBacked: Boolean = false,
    val amount: Int = TransferOwnershipSampleData.DEFAULT_AMOUNT,
    val confirmationField: FormFieldState = FormFieldState(id = "confirmation"),
    val sheetPhase: ConfirmSheetPhase = ConfirmSheetPhase.Hidden,
    val biometricErrorMessage: String? = null,
    val toast: TransferToast? = null,
    val shouldDismiss: Boolean = false,
    val biometryLabel: String = "Fingerprint",
) {
    val maxAmount: Int get() = currentUser.percent
    val presets: List<Int> get() = TransferOwnershipSampleData.presets
    val sliderRange: IntRange get() = TransferOwnershipSampleData.sliderRange
    val confirmationPhrase: String get() = TransferOwnershipSampleData.CONFIRMATION_PHRASE

    val confirmationMatches: Boolean
        get() = confirmationField.value == confirmationPhrase

    val isReadyToCommit: Boolean
        get() = amount in 1..maxAmount && confirmationMatches && recipientIsBackendBacked

    val isDirty: Boolean
        get() = amount != TransferOwnershipSampleData.DEFAULT_AMOUNT || confirmationField.value.isNotEmpty()

    val confirmationFieldState: PantopusFieldState
        get() =
            when {
                confirmationField.value.isEmpty() -> PantopusFieldState.Default
                confirmationMatches -> PantopusFieldState.Valid
                else -> PantopusFieldState.Default
            }

    val ctaLabel: String
        get() {
            if (!recipientIsBackendBacked) return "Transfer ownership unavailable"
            val firstName = recipient.name.split(" ").firstOrNull() ?: recipient.name
            return "Transfer $amount% to $firstName"
        }

    val setupBlockMessage: String?
        get() =
            if (recipientIsBackendBacked) {
                null
            } else {
                "Recipient selection is not connected to live owner data yet."
            }

    val warningCopy: String
        get() {
            val firstName = recipient.name.split(" ").firstOrNull() ?: recipient.name
            val others = coOwners.joinToString(separator = " and ") { it.displayName }
            return "$others will be notified after this transfer. You cannot reclaim the $amount% " +
                "without $firstName's signed transfer back."
        }

    val confirmationTimestamp: String
        get() = "14:23 May 26"

    val beforeSegments: List<SplitSegment>
        get() {
            val you =
                SplitSegment(
                    id = currentUser.id,
                    owner = currentUser.displayName,
                    percent = currentUser.percent,
                    color = currentUser.palette.color,
                )
            val others =
                coOwners.map { owner ->
                    SplitSegment(
                        id = owner.id,
                        owner = owner.displayName,
                        percent = owner.percent,
                        color = owner.palette.color,
                    )
                }
            return listOf(you) + others
        }

    val afterSegments: List<SplitSegment>
        get() {
            val recipientFirstName = recipient.name.split(" ").firstOrNull() ?: recipient.name
            val you =
                SplitSegment(
                    id = currentUser.id,
                    owner = currentUser.displayName,
                    percent = (currentUser.percent - amount).coerceAtLeast(0),
                    color = currentUser.palette.color,
                    delta = -amount,
                )
            val newcomer =
                SplitSegment(
                    id = recipient.id,
                    owner = recipientFirstName,
                    percent = amount,
                    color = TransferOwnershipSampleData.recipientPaletteStart,
                    delta = amount,
                    isNew = true,
                )
            val others =
                coOwners.map { owner ->
                    SplitSegment(
                        id = owner.id,
                        owner = owner.displayName,
                        percent = owner.percent,
                        color = owner.palette.color,
                    )
                }
            return listOf(you, newcomer) + others
        }

    val confirmSheetParties: List<ConfirmSheetParty>
        get() =
            listOf(
                ConfirmSheetParty(
                    id = currentUser.id,
                    role = "From",
                    name = "You · ${TransferOwnershipSampleData.SENDER_FULL_NAME}",
                    initials = currentUser.initials,
                    avatarStart = currentUser.palette.gradientStart,
                    avatarEnd = currentUser.palette.gradientEnd,
                    fromPercent = currentUser.percent,
                    toPercent = (currentUser.percent - amount).coerceAtLeast(0),
                ),
                ConfirmSheetParty(
                    id = recipient.id,
                    role = "To",
                    name = recipient.name,
                    initials = recipient.initials,
                    avatarStart = TransferOwnershipSampleData.recipientPaletteStart,
                    avatarEnd = TransferOwnershipSampleData.recipientPaletteEnd,
                    fromPercent = 0,
                    toPercent = amount,
                    verified = recipient.verified,
                ),
            )
}

/**
 * A13.4 — Transfer Ownership form view-model. After biometric auth
 * succeeds, [handleBiometricResult] calls
 * `POST /api/homes/:id/owners/transfer` (route
 * `backend/routes/homeOwnership.js:1526`) via [HomeOwnersRepository],
 * raises a success toast (including the co-owner-approval quorum
 * branch), and signals [shouldDismiss].
 *
 * Biometric authentication itself is owned by the host screen — the VM
 * exposes [requestBiometric] / [handleBiometricResult] entry points so
 * the platform [`androidx.biometric.BiometricPrompt`] lives at the
 * Activity layer where it has the FragmentManager it needs.
 */
@HiltViewModel
class TransferOwnershipViewModel
    @Inject
    constructor(
        savedStateHandle: SavedStateHandle,
        private val ownersRepo: HomeOwnersRepository,
    ) : ViewModel() {
        private val homeId: String = savedStateHandle.get<String>(TRANSFER_HOME_ID_KEY) ?: ""

        private val _state =
            MutableStateFlow(
                TransferOwnershipUiState(
                    homeContext = TransferOwnershipSampleData.homeContext(homeId),
                    recipient = TransferOwnershipSampleData.mayaFortune,
                    currentUser = TransferOwnershipSampleData.currentUser,
                    coOwners = TransferOwnershipSampleData.coOwners,
                ),
            )
        val state: StateFlow<TransferOwnershipUiState> = _state.asStateFlow()

        internal constructor(
            savedStateHandle: SavedStateHandle,
            ownersRepo: HomeOwnersRepository,
            recipientIsBackendBacked: Boolean,
        ) : this(savedStateHandle, ownersRepo) {
            _state.update { it.copy(recipientIsBackendBacked = recipientIsBackendBacked) }
        }

        fun updateAmount(raw: Int) {
            _state.update { current ->
                val clamped = raw.coerceIn(current.sliderRange.first, current.maxAmount)
                current.copy(amount = clamped)
            }
        }

        fun selectPreset(preset: Int) {
            updateAmount(preset)
        }

        fun updateConfirmation(value: String) {
            _state.update { current ->
                current.copy(confirmationField = current.confirmationField.copy(value = value, touched = true))
            }
        }

        fun presentConfirmSheet() {
            val current = _state.value
            val setupBlockMessage = current.setupBlockMessage
            if (setupBlockMessage != null) {
                _state.update {
                    it.copy(
                        toast = TransferToast(text = setupBlockMessage, isError = true),
                    )
                }
                return
            }
            if (!current.isReadyToCommit) return
            _state.update { it.copy(sheetPhase = ConfirmSheetPhase.Visible, biometricErrorMessage = null) }
        }

        fun dismissConfirmSheet() {
            val current = _state.value
            if (current.sheetPhase == ConfirmSheetPhase.Authenticating) return
            _state.update { it.copy(sheetPhase = ConfirmSheetPhase.Hidden, biometricErrorMessage = null) }
        }

        /**
         * Marks authentication as in-flight. The host calls this just
         * before invoking the platform BiometricPrompt.
         */
        fun requestBiometric() {
            val current = _state.value
            if (current.sheetPhase != ConfirmSheetPhase.Visible || !current.isReadyToCommit) return
            _state.update { it.copy(sheetPhase = ConfirmSheetPhase.Authenticating, biometricErrorMessage = null) }
        }

        /**
         * Drive the post-auth state machine. On success runs the stub
         * transfer; on failure surfaces an inline error and returns the
         * sheet to the Visible phase so the user can retry.
         */
        fun handleBiometricResult(
            success: Boolean,
            errorMessage: String? = null,
        ) {
            if (!success) {
                _state.update {
                    it.copy(
                        sheetPhase = ConfirmSheetPhase.Visible,
                        biometricErrorMessage = errorMessage ?: "Authentication failed. Try again.",
                    )
                }
                return
            }
            viewModelScope.launch {
                val current = _state.value
                val setupBlockMessage = current.setupBlockMessage
                if (setupBlockMessage != null) {
                    _state.update {
                        it.copy(
                            sheetPhase = ConfirmSheetPhase.Visible,
                            biometricErrorMessage = setupBlockMessage,
                        )
                    }
                    return@launch
                }
                // Identify the buyer by their account id; effective_date
                // is omitted so the transfer takes effect immediately.
                val result =
                    ownersRepo.transfer(
                        homeId,
                        TransferOwnerRequest(buyerUserId = current.recipient.id),
                    )
                when (result) {
                    is NetworkResult.Success -> {
                        val successText =
                            if (result.data.requiresQuorum) {
                                "Transfer sent for co-owner approval."
                            } else {
                                "Transferred ${current.amount}% to ${current.recipient.name}"
                            }
                        _state.update {
                            it.copy(
                                sheetPhase = ConfirmSheetPhase.Dismissing,
                                toast = TransferToast(text = successText, isError = false),
                                shouldDismiss = true,
                            )
                        }
                    }
                    is NetworkResult.Failure -> {
                        _state.update {
                            it.copy(
                                sheetPhase = ConfirmSheetPhase.Visible,
                                biometricErrorMessage =
                                    result.error.message.ifEmpty {
                                        "Couldn't complete the transfer. Try again."
                                    },
                            )
                        }
                    }
                }
            }
        }

        fun setBiometryLabel(label: String) {
            _state.update { it.copy(biometryLabel = label) }
        }

        fun dismissToast() {
            _state.update { it.copy(toast = null) }
        }

        fun acknowledgeDismiss() {
            _state.update { it.copy(shouldDismiss = false) }
        }
    }

File: frontend/apps/ios/Pantopus/Features/Homes/Owners/Transfer/TransferOwnershipViewModel.swift
//
//  TransferOwnershipViewModel.swift
//  Pantopus
//
//  A13.4 — Backs the Transfer Ownership form. Holds the recipient
//  selection, the live transfer amount (1–60% slider), the typed
//  confirmation field, the bottom-sheet visibility, and the Face ID
//  authentication state machine. The current recipient projection is
//  sample-only, so the live app blocks the backend mutation until recipient
//  selection is backed by real owner data. Tests can opt into the transfer
//  executor seam to keep the post-auth state machine covered.
//

import Foundation
import LocalAuthentication
import Observation
import SwiftUI

/// Visibility states for the bottom confirmation sheet.
public enum ConfirmSheetPhase: Sendable, Equatable {
    case hidden
    case visible
    case authenticating
    case dismissing
}

@Observable
@MainActor
public final class TransferOwnershipViewModel {
    // MARK: - Inputs

    public let homeId: String
    public let homeContext: TransferOwnershipSampleData.HomeContext
    public let recipient: TransferOwnershipSampleData.RecipientSeed
    public let currentUser: TransferOwnershipSampleData.OwnerSeed
    public let coOwners: [TransferOwnershipSampleData.OwnerSeed]
    public let recipientIsBackendBacked: Bool
    public let presets: [Int] = TransferOwnershipSampleData.presets
    public let sliderRange: ClosedRange<Int> = TransferOwnershipSampleData.sliderRange
    public let confirmationPhrase: String = TransferOwnershipSampleData.confirmationPhrase

    // MARK: - Mutable state

    public var amount: Int
    public var confirmationField: FormFieldState
    public private(set) var sheetPhase: ConfirmSheetPhase = .hidden
    public private(set) var biometricErrorMessage: String?
    public var toast: ToastMessage?
    public private(set) var shouldDismiss = false

    // MARK: - Injected boundary

    /// Biometric evaluator. Default uses `LocalAuthentication`; tests can
    /// inject a deterministic stub that returns `.success` / `.failure`
    /// without prompting the user.
    public typealias BiometricEvaluator = @MainActor (_ reason: String) async -> Result<Void, any Error>
    private let biometricEvaluator: BiometricEvaluator

    /// Test seam for the backend round-trip. When `nil` (the default in
    /// the live app) `commitTransfer()` performs
    /// `POST /api/homes/:id/owners/transfer` through `api`; tests inject
    /// a deterministic stub so they never touch the network.
    public typealias TransferExecutor = @MainActor () async throws -> Void
    private let transferExecutor: TransferExecutor?
    private let api: APIClient

    // MARK: - Init

    init(
        homeId: String,
        amount: Int = TransferOwnershipSampleData.defaultAmount,
        api: APIClient = .shared,
        biometricEvaluator: BiometricEvaluator? = nil,
        transferExecutor: TransferExecutor? = nil,
        recipientIsBackendBacked: Bool = false
    ) {
        self.homeId = homeId
        self.api = api
        self.recipientIsBackendBacked = recipientIsBackendBacked
        homeContext = TransferOwnershipSampleData.homeContext(for: homeId)
        recipient = TransferOwnershipSampleData.mayaFortune
        currentUser = TransferOwnershipSampleData.currentUser
        coOwners = TransferOwnershipSampleData.coOwners
        self.amount = max(1, min(amount, currentUser.percent))
        confirmationField = FormFieldState(id: "confirmation", originalValue: "")
        self.biometricEvaluator = biometricEvaluator ?? Self.defaultBiometricEvaluator
        self.transferExecutor = transferExecutor
    }

    // MARK: - Computed projections

    public var maxAmount: Int {
        currentUser.percent
    }

    /// Whether the typed phrase matches the literal "TRANSFER" exactly.
    public var confirmationMatches: Bool {
        confirmationField.value == confirmationPhrase
    }

    /// Whether the sticky CTA is active.
    public var isReadyToCommit: Bool {
        amount > 0 && amount <= maxAmount && confirmationMatches && recipientIsBackendBacked
    }

    /// Whether the host should arm the dirty-close confirm. Any input
    /// touched flips this on.
    public var isDirty: Bool {
        amount != TransferOwnershipSampleData.defaultAmount
            || !confirmationField.value.isEmpty
    }

    /// Field state visual for the typed confirmation input.
    public var confirmationFieldState: PantopusFieldState {
        if confirmationField.value.isEmpty { return .default }
        return confirmationMatches ? .valid : .default
    }

    public var ctaLabel: String {
        if !recipientIsBackendBacked { return "Transfer ownership unavailable" }
        return "Transfer \(amount)% to \(recipient.name.split(separator: " ").first.map(String.init) ?? recipient.name)"
    }

    public var setupBlockMessage: String? {
        recipientIsBackendBacked ? nil : "Recipient selection is not connected to live owner data yet."
    }

    public var warningCopy: String {
        let firstName = recipient.name.split(separator: " ").first.map(String.init) ?? recipient.name
        return "\(coOwners.map(\.displayName).joined(separator: " and ")) " +
            "will be notified after this transfer. You cannot reclaim the \(amount)% without " +
            "\(firstName)'s signed transfer back."
    }

    public var biometryLabel: String {
        Self.biometryLabel
    }

    // MARK: - Diff projection

    public var beforeSegments: [SplitSegment] {
        let you = SplitSegment(
            id: currentUser.id,
            owner: currentUser.displayName,
            percent: currentUser.percent,
            color: currentUser.palette.color
        )
        let others = coOwners.map { owner in
            SplitSegment(
                id: owner.id,
                owner: owner.displayName,
                percent: owner.percent,
                color: owner.palette.color
            )
        }
        return [you] + others
    }

    public var afterSegments: [SplitSegment] {
        let recipientFirstName = recipient.name.split(separator: " ").first.map(String.init) ?? recipient.name
        let you = SplitSegment(
            id: currentUser.id,
            owner: currentUser.displayName,
            percent: max(0, currentUser.percent - amount),
            color: currentUser.palette.color,
            delta: -amount
        )
        let newcomer = SplitSegment(
            id: recipient.id,
            owner: recipientFirstName,
            percent: amount,
            color: TransferOwnershipSampleData.recipientPaletteStart,
            delta: amount,
            isNew: true
        )
        let others = coOwners.map { owner in
            SplitSegment(
                id: owner.id,
                owner: owner.displayName,
                percent: owner.percent,
                color: owner.palette.color
            )
        }
        return [you, newcomer] + others
    }

    public var confirmSheetParties: [ConfirmSheetParty] {
        [
            ConfirmSheetParty(
                id: currentUser.id,
                role: "From",
                name: "You · \(TransferOwnershipSampleData.senderFullName)",
                initials: currentUser.initials,
                avatarStart: currentUser.palette.gradientStart,
                avatarEnd: currentUser.palette.gradientEnd,
                fromPercent: currentUser.percent,
                toPercent: max(0, currentUser.percent - amount)
            ),
            ConfirmSheetParty(
                id: recipient.id,
                role: "To",
                name: recipient.name,
                initials: recipient.initials,
                avatarStart: TransferOwnershipSampleData.recipientPaletteStart,
                avatarEnd: TransferOwnershipSampleData.recipientPaletteEnd,
                fromPercent: 0,
                toPercent: amount,
                verified: recipient.verified
            )
        ]
    }

    /// Stable "HH:mm MMM d" stamp shown in the legal copy. Hardcoded for
    /// snapshot reproducibility.
    public var confirmationTimestamp: String {
        "14:23 May 26"
    }

    // MARK: - Mutations

    public func updateAmount(_ raw: Int) {
        amount = min(max(sliderRange.lowerBound, raw), maxAmount)
    }

    public func selectPreset(_ preset: Int) {
        updateAmount(preset)
    }

    public func updateConfirmation(_ value: String) {
        confirmationField.value = value
        confirmationField.touched = true
    }

    public func presentConfirmSheet() {
        if let setupBlockMessage {
            toast = ToastMessage(text: setupBlockMessage, kind: .error)
            return
        }
        guard isReadyToCommit else { return }
        sheetPhase = .visible
        biometricErrorMessage = nil
    }

    public func dismissConfirmSheet() {
        guard sheetPhase != .authenticating else { return }
        sheetPhase = .hidden
        biometricErrorMessage = nil
    }

    /// Authenticate with biometrics, then run the stubbed transfer
    /// endpoint. On success raises a toast and signals the host should
    /// dismiss. On biometric failure the sheet stays open and surfaces an
    /// inline error so the user can retry.
    public func authenticateAndCommit() async {
        guard sheetPhase == .visible, isReadyToCommit else { return }
        sheetPhase = .authenticating
        biometricErrorMessage = nil
        let reason = "Confirm transfer of \(amount)% to \(recipient.name)"
        let result = await biometricEvaluator(reason)
        switch result {
        case .success:
            await commitTransfer()
        case let .failure(error):
            sheetPhase = .visible
            biometricErrorMessage = (error as? LAError)
                .map { Self.message(for: $0) }
                ?? error.localizedDescription
        }
    }

    public func acknowledgeDismiss() {
        shouldDismiss = false
    }

    private func commitTransfer() async {
        if let setupBlockMessage {
            sheetPhase = .visible
            biometricErrorMessage = setupBlockMessage
            return
        }
        do {
            let successText: String
            if let transferExecutor {
                // Test seam — deterministic stub, no network.
                try await transferExecutor()
                successText = "Transferred \(amount)% to \(recipient.name)"
            } else {
                // Identify the buyer by their account id. `effective_date`
                // is omitted so the transfer takes effect immediately.
                let request = TransferOwnerRequest(buyerUserId: recipient.id)
                let response = try await api.request(
                    HomesEndpoints.transferOwner(homeId: homeId, request: request),
                    as: TransferOwnerResponse.self
                )
                successText = response.requiresQuorum
                    ? "Transfer sent for co-owner approval."
                    : "Transferred \(amount)% to \(recipient.name)"
            }
            sheetPhase = .dismissing
            toast = ToastMessage(text: successText, kind: .success)
            shouldDismiss = true
        } catch {
            sheetPhase = .visible
            biometricErrorMessage = (error as? APIError)?.errorDescription
                ?? "Couldn't complete the transfer. Try again."
        }
    }

    // MARK: - Defaults

    private static let defaultBiometricEvaluator: BiometricEvaluator = { reason in
        let context = LAContext()
        var error: NSError?
        let policy: LAPolicy = .deviceOwnerAuthentication // biometrics with passcode fallback
        guard context.canEvaluatePolicy(policy, error: &error) else {
            return .failure(error ?? LAError(.biometryNotAvailable))
        }
        return await withCheckedContinuation { continuation in
            context.evaluatePolicy(policy, localizedReason: reason) { success, evalError in
                if success {
                    continuation.resume(returning: .success(()))
                } else {
                    continuation.resume(returning: .failure(evalError ?? LAError(.authenticationFailed)))
                }
            }
        }
    }

    private static var biometryLabel: String {
        let context = LAContext()
        _ = context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: nil)
        switch context.biometryType {
        case .faceID: return "Face ID"
        case .touchID: return "Touch ID"
        case .opticID: return "Optic ID"
        case .none: return "Passcode"
        @unknown default: return "Face ID"
        }
    }

    private static func message(for error: LAError) -> String {
        switch error.code {
        case .userCancel, .systemCancel, .appCancel:
            "Authentication was cancelled."
        case .authenticationFailed:
            "Authentication failed. Try again."
        case .passcodeNotSet:
            "Set a device passcode to confirm transfers."
        case .biometryNotAvailable, .biometryNotEnrolled, .biometryLockout:
            "Biometric authentication isn't available right now."
        default:
            error.localizedDescription
        }
    }
}
